src/Functions/FunctionsHashing.h

```diff
@@ -413,23 +414,48 @@ struct MurmurHash3Impl128
     static constexpr bool use_int_hash_for_pods = false;
 };
 
-/// http://hg.openjdk.java.net/jdk8u/jdk8u/jdk/file/478a4add975b/src/share/classes/java/lang/String.java#l1452
 /// Care should be taken to do all calculation in unsigned integers (to avoid undefined behaviour on overflow)
 ///  but obtain the same result as it is done in signed integers with two's complement arithmetic.
 struct JavaHashImpl
 {
     static constexpr auto name = "javaHash";
     using ReturnType = Int32;
 
-    static Int32 apply(const char * data, const size_t size)
+    static ReturnType apply(int64_t x)
+    {
+        int64_t copy = x;
+        copy = copy >> 32;
```

Comment:
This code is not correct in C++:

> For signed and non-negative a, if a * 2b
is representable in the unsigned version of the return type, then that value, [converted](https://en.cppreference.com/w/cpp/language/implicit_conversion#Integral_conversions) to signed, is the value of a << b (this makes it legal to create [INT_MIN](https://en.cppreference.com/w/cpp/types/climits) as 1 << 31); otherwise the behavior is undefined.

> For negative a, the behavior of a << b is undefined.

https://en.cppreference.com/w/cpp/language/operator_arithmetic

Reply:
A random page on the internet is saying that 
> `>>>` is an unsigned right shift operator

that is different from your code.

Reply:
1.  According the doc, Since C++20 `>>` is arithmetic right shift
> Since C++20
> The value of a >> b is a/2b, rounded down (in other words, right shift on signed a is arithmetic right shift).

2.  `>>>` is implemented by  `copy >> 32`  and  `copy & 0x00000000FFFFFFFF`

Reply:
Ok.